beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -2971,6 +2981,22 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                 info!(self.log, "Writing blobs to store"; "block_root" => ?block_root);
                 ops.push(StoreOp::PutBlobs(block_root, blobs));
             }
+
+            // Update db's metadata for blobs pruning.
+            if current_slot == current_epoch.start_slot(T::EthSpec::slots_per_epoch()) {
+                if let Some(mut blob_info) = self.store.get_blob_info() {
+                    let next_epoch_to_prune =
+                        blob_info.last_pruned_epoch + *MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS;
+
+                    if current_epoch > next_epoch_to_prune {
+                        blob_info.data_availability_breakpoint = Some(block_root);
+                        self.store.compare_and_set_blob_info_with_write(
+                            self.store.get_blob_info(),
```

Comment:
Wouldn't it be better to take a write lock on blob info then before reading `orig_blob_info` because another thread could edit another field on `BlobInfo` before we get to `compare_and_set(orig_blob_info, new_blob_info)`?

Reply:
Potentially. If we're reading the `blob_info` a lot but only writing it from one spot when pruning then it could be better to avoid holding the write lock. In that case the compare-and-swap acts as a runtime safeguard for the structural property that only 1 thread should be attempting to write at a time.

However, at the moment we're starting pruning on every new head block _and_ from the finalization migrator, so there could be multiple pruning trying jobs to run in parallel (which would block if we held the lock, or fail the strict version of compare-and-swap).

We probably want to parallel pruning jobs altogether, which we could do by sending the pruning jobs on a channel to a background thread (like the migrator thread). We could even just re-use the existing migrator thread by adding a notification type here for blob pruning:

https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/beacon_chain/src/migrate.rs#L85-L89

Reply:
Ok, I think I have took into account all you have explained in your review and changed the code. The problem of blocking I removed by simplifying the conceptual design, like moving all code that changes BlobInfo into `try_prune_blobs` and spawning blob pruning into a thread (except in the `database_manager` because that's run offline right so doesn't interfere with node business?). Also decided on one place to call try_prune blobs in the finalisation process, one which to me makes the most sense. Let me know what you think.

Reply:
> Also decided on one place to call try_prune_blobs in the finalisation process, one which to me makes the most sense

So the issue with this is in periods where finalization doesn't advance, we won't prune blobs at all. And finalization not advancing is a symptom of general bad network health so adding unbounded blob db growth would be bad.

This sounds like a good approach to me:
> We probably want to parallel pruning jobs altogether, which we could do by sending the pruning jobs on a channel to a background thread (like the migrator thread). We could even just re-use the existing migrator thread by adding a notification type here for blob pruning:

I'm not sure exactly what should trigger the blob prune notification. Maybe since it's purely time based, it could just be sent from the `per_slot_task` every so often

Reply:
Aha, ok I see, so this is why Michael said `after_new_head`! My bad, lost that info along the way. Maybe `per_slot_task` is a better place to send a `Notification::PruneBlobs` to the migrator thread than `after_new_head, since the future may entail a data availability boundary before finalisation. What do you think? And hints on how to prune head candidates based on the slot they want to fill?

Reply:
`after_new_head` also means that the canonical head has advanced right? So if `after_new_head` is called, `migrate_database` should be called and update the `split.slot` if there isn't a bug on the way right?

Reply:
@sean based on [this comment](https://github.com/sigp/lighthouse/pull/3891#discussion_r1072401598) about making a marginal for blobs pruning, I think my logic checks out: falling back to using the `split.slot` to calculate current epoch if the prune is manual (using database manager) and there is no access to the current `data_availability_boundary`, and otherwise, if the prune is automatically triggered by finalization, the function `data_availability_boundary` is called and passed to `try_prune_blobs`. 

If on the contrary you would say that pruning should always happen exactly on the data availability boundary, then I would favour storing an `Arc<BeaconChain<T>>` to the beacon chain in the migrator and this way getting the current epoch on both manual and automatic prune. But even in the future, if the data availability boundary is less than the finalisation epoch, if a marginal of 2 epochs is alright then we can still keep my logic. If we do want the pruning to be exactly on the data availability boundary and that boundary is earlier than finalisation, the migrator might be the best place to store an `Arc<HeadTracker>` as the combo manual+automatic prune doesn't allow us to pass that info in a Notification like we do for `FinalizationNotification` (no access to it via manual prune).

I think, the question that remains is: do we want to spawn the automatic pruning in a random thread or use the migrator thread?

Reply:
> @sean based on https://github.com/sigp/lighthouse/pull/3891#discussion_r1072401598 about making a marginal for blobs pruning, I think my logic checks out: falling back to using the split.slot to calculate current epoch if the prune is manual (using database manager) and there is no access to the current data_availability_boundary, and otherwise, if the prune is automatically triggered by smthg, the function data_availability_boundary is called and passed to try_prune_blobs.

I think passing in the boundary epoch is the right way to go, had some additional comments about it [here](https://github.com/sigp/lighthouse/pull/3852#discussion_r1073768743). 

> do we want to spawn the automatic pruning in a random thread or use the migrator thread?

The migrator thread sounds like a good spot to me

> Where in per_slot_task should blobs pruning be triggered?

If we use `per_slot_task` we could just check if `slot % EPOCHS_PER_BLOB_PUNE == 0` and send a message to the migrator, then we could let the migrator figure out when exactly during the slot the prune should be run. We might as well but this at the end of the method because it runs some other important operations first. I noticed the migrator can be configured to block, but I don't see that being used anywhere. If we need to fallback to a blocking version of this prune, we should consider not using `per_slot_task` at all.  Maybe `after_new_head` would be a better spot if that's the case.  What're your thoughts @michaelsproul 
 
> How flexible can the margin be, should we adapt the code now to be suitable for pruning pre-finalisation blobs?

If we want to add a margin to the blob prune somewhere we can add something like`const BLOB_PUNE_MARGIN_EPOCHS : u64 = 10` that we subtract from from the data availability boundary we pass into `try_prune_blobs`, we could make this configurable too (probably not worth it for now). I don't think we should yet include pre-finalized blob pruning because it seems like it'll add a lot of complexity. and requires ~20 days of non-finalization. I think it's a reasonable next-step type of task

Reply:
I didn't realise that the canonical head is not always finalised. Love learning at the deep end, it's how the knowledge really sticks, when the mistake hurts my ego! lmao. If we aren't getting a new head we are most likely not importing many more new blocks&blobs so we aren't getting a blobs state bloat, so `after_new_head` sounds like a good place to me too now.